### PR TITLE
Finish `MerkleMMCS` Implementation

### PR DESCRIPTION
**Pull request by Sladuca**
Addresses #70.

Summary of changes:
- added doc comments to `MMCS` interface
- changed `verify_batch` function signature to take a slice instead of an owned `Vec` (discussed with @dlubarov offline)
- added `PartialEq` bound for `D` in `MerkleMMCS` (needed to check roots match in `verify_batch`)
- made bounds consistent for `D` between `MerkleMMCS`, `MerkleTree`, and impls
- finished implementation of `open_batch`
- finished implementation of `verify_batch`
- added height property check to `MerkleTree::new`
- added more tests to `MerkleMMCS`

### CODE REVIEW

File: merkle-tree/src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#![no_std]
```
Comment (dlubarov) [inline merkle-tree/src/lib.rs:1]: Was this just for debugging? Would prefer to keep no_std

File: merkle-tree/src/lib.rs [hunk 2]
```diff
@@ -35,13 +33,23 @@ impl<L, D> MerkleTree<L, D> {
...
     {
         assert!(!leaves.is_empty(), "No matrices given?");
 
-        // TODO: Check the matching height condition.
+        // check height property
+        assert!(
+            leaves
+                .iter()
+                .map(|m| m.height())
+                .sorted_by_key(|&h| Reverse(h))
```
Comment (dlubarov) [inline merkle-tree/src/lib.rs:47]: nit: could this be `sorted()` with no key? Seems like reversing should be optional

File: merkle-tree/src/lib.rs [hunk 3]
```diff
@@ -181,20 +189,24 @@ where
...
-                let log2_height = log2_strict_usize(matrix.height());
+                let log2_height = log2_ceil_usize(matrix.height());
                 let bits_reduced = log_max_height - log2_height;
                 let reduced_index = index >> bits_reduced;
                 matrix.row(reduced_index).collect()
             })
+            .collect_vec();
+
+        let proof = (0..log_max_height)
+            .map(|i| prover_data.digest_layers[i][(index >> i) ^ 1])
```
Comment (dlubarov) [inline merkle-tree/src/lib.rs:206]: Nice and clean 😎

File: merkle-tree/src/lib.rs (hunk 1 above)
Comment (Sladuca) [reply merkle-tree/src/lib.rs:1]: yes it was, sorry about that

**Review by dlubarov — APPROVED** [review]
Looks great! 👍

**Pull request merged**